File: js/src/services/LogParserService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogParserService = void 0;
const AreaChangeParser_1 = require("../parsers/AreaChangeParser");
const PlayerJoinedAreaParser_1 = require("../parsers/PlayerJoinedAreaParser");
const GlobalMessageParser_1 = require("../parsers/GlobalMessageParser");
const IncomingTradeParser_1 = require("../parsers/IncomingTradeParser");
const OutgoingTradeParser_1 = require("../parsers/OutgoingTradeParser");
const LogEntry_1 = require("../models/entries/LogEntry");
const LogEntryTag_1 = require("../enums/LogEntryTag");
const LogEntryType_1 = require("../enums/LogEntryType");
class LogParserService {
    constructor() {
        this._regLogEntryTag = /\[(debug|warn|info) Client [0-9]+\] /gi;
        this._parsers = [
            new AreaChangeParser_1.AreaChangeParser(),
            new PlayerJoinedAreaParser_1.PlayerJoinedAreaParser(),
            new GlobalMessageParser_1.GlobalMessageParser(),
            new IncomingTradeParser_1.IncomingTradeParser(),
            new OutgoingTradeParser_1.OutgoingTradeParser()
        ];
    }
    parse(line) {
        let entry = this.parseLogEntry(line);
        if (!entry)
            return;
        for (let parser of this._parsers.filter(p => p.canParse(line))) {
            try {
                return parser.execute(entry.entry, entry.line);
            }
            catch (_a) { }
        }
        entry.entry.types = [LogEntryType_1.LogEntryType.System];
        return entry.entry;
    }
    addParser(parser) {
        this._parsers.push(parser);
    }
    removeAllParsers() {
        this._parsers = [];
    }
    parseLogEntry(line) {
        var entry = new LogEntry_1.LogEntry(line, new Date(), [], LogEntryTag_1.LogEntryTag.Info);
        var dateEndIndex = line.indexOf(" ");
        if (dateEndIndex == -1) {
            return;
        }
        var timeEndIndex = line.indexOf(" ", dateEndIndex);
        if (timeEndIndex == -1) {
            return;
        }
        var dateTimeEndIndex = (dateEndIndex + timeEndIndex) - 1;
        var timeStr = line.substring(0, dateTimeEndIndex);
        line = line.substring((dateTimeEndIndex + 1));
        var logTagMatch = this._regLogEntryTag.exec(line);
        if (!logTagMatch) {
            return;
        }
        var tagStr = line.substring(logTagMatch.index + 1, line.indexOf(" ", logTagMatch.index));
        tagStr = tagStr.substring(0, tagStr.indexOf(" "));
        line = line.substring(logTagMatch.index + logTagMatch.length);
        entry.time = new Date(timeStr);
        entry.tag = LogEntryTag_1.ConvertLogEntryTag(tagStr);
        return { entry, line };
    }
}
exports.LogParserService = LogParserService;

File: dotnet/PoeLogsParser/Services/LogParserService.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using PoeLogsParser.Enums;
using PoeLogsParser.Models;
using PoeLogsParser.Models.Abstractions;
using PoeLogsParser.Parsers;
using PoeLogsParser.Parsers.Abstractions;
using PoeLogsParser.Services.Abstractions;

namespace PoeLogsParser.Services
{
    public class LogParserService : ILogParserService
    {
        private readonly CurrencyService _currencyService;
        private List<IParser> _parsers;

        private readonly Regex _regLogEntryTag =
            new Regex("\\[(debug|warn|info) Client [0-9]+\\] ", RegexOptions.IgnoreCase);

        public LogParserService()
        {
            _currencyService = new CurrencyService();
            _parsers = new List<IParser>()
            {
                new AreaChangeParser(),
                new PlayerJoinedAreaParser(),
                new TradeChatMessageParser(),
                new GlobalMessageParser(),
                new IncomingTradeParser(),
                new OutgoingTradeParser(),
                new TradeStateParser()
            };
        }

        public ILogEntry Parse(string line)
        {
            var entry = ParseLogEntry(line);

            if (entry == null)
            {
                return default(LogEntry);
            }

            foreach (var parser in _parsers.Where(parser => parser.CanParse(line)))
            {
                try
                {
                    return parser.Execute(entry.Item2, entry.Item1);
                } catch
                {
                    // ignored
                }
            }

            entry.Item1.Types = new[] {LogEntryType.System};

            return entry.Item1;
        }

        private Tuple<LogEntry, string> ParseLogEntry(string line)
        {
            var entry = new LogEntry(line, DateTime.Now, new List<LogEntryType>());

            var dateEndIndex = line.IndexOf(" ", StringComparison.Ordinal);

            if (dateEndIndex == -1)
            {
                return default;
            }

            var timeEndIndex = line.IndexOf(" ", dateEndIndex, StringComparison.Ordinal);

            if (timeEndIndex == -1)
            {
                return default;
            }

            var dateTimeEndIndex = (dateEndIndex + timeEndIndex) - 1;

            if(dateTimeEndIndex == -1) return default;

            var timeStr = line[..dateTimeEndIndex];
            line = line[(dateTimeEndIndex + 1)..];

            var logTagMatch = _regLogEntryTag.Match(line);

            if (!logTagMatch.Success)
            {
                return default;
            }

            var tagStr = line.Substring(logTagMatch.Index + 1,
                line.IndexOf(" ", logTagMatch.Index, StringComparison.Ordinal));
            tagStr = tagStr[..tagStr.IndexOf(" ", StringComparison.Ordinal)];
            line = line[(logTagMatch.Index + logTagMatch.Length)..];

            entry.Time = DateTime.Parse(timeStr);
            entry.Tag = LogEntryTagConveter.Convert(tagStr);

            return new Tuple<LogEntry, string>(entry, line);
        }

        public void AddParser(IParser parser)
        {
            _parsers.Add(parser);
        }

        public void RemoveAllParsers()
        {
            _parsers.Clear();
        }
    }
}